libqtile/core/manager.py

```diff
@@ -1341,7 +1347,12 @@ def spawn(
                 file_actions.append((os.POSIX_SPAWN_CLOSEFROM, 3))
 
             try:
-                return os.posix_spawnp(args[0], args, env, file_actions=file_actions)
+                pid = os.posix_spawnp(args[0], args, env, file_actions=file_actions)
+                if group is not None:
```

Comment:
This is always going to be slightly racy: if the pid launches and creates a window before we add the pid to the rules list, we won't catch it.

To fix that, I think the following would work:

1. add a `qtile launch --fd N ...` which waits for a character to be sent on FD N, and then `exec()s` the process
2. changes this to `posix_spawnp()` that doesn't close all the fds, and prepends `qtile launch --fd N` to that
3. in this code, we can create a socket pair, launch the `qtile launch` process, add that pid rule, send the notification on the socket pair

Since the `exec()` from `qtile launch` will keep the same pid, the rule match should work.

Reply:
We've got `qtile run-cmd` which is where this `Rule`-based approach came from. I was already suspicious about the race condition there but it seemed to work well and this code also worked well in (admittedly, limited) practice but I wasn't surprised to see it fail in CI.

I *think* I understand your steps above (I'm still relatively new to sockets/file descriptors) and will give it a try.

Reply:
Oh yeah, I wonder if we have tests for that. But in any case, that's racy too.